Remove depprojs in favor of PackageReferences

ViktorHofer opened the pull request:
Fixes #31844
Fixes https://github.com/dotnet/runtime/issues/31872
Fixes https://github.com/dotnet/runtime/issues/963
Fixes https://github.com/dotnet/runtime/issues/37959
Fixes https://github.com/dotnet/runtime/issues/23972
Fixes https://github.com/dotnet/runtime/issues/35061
Fixes https://github.com/dotnet/runtime/issues/1919
Fixes https://github.com/dotnet/runtime/issues/31757
Fixes https://github.com/dotnet/runtime/issues/38278
Works towards https://github.com/dotnet/runtime/issues/28244
Requires https://github.com/dotnet/arcade/pull/5575

## Changes

- Remove depprojs which currently binplace external references into the RefPath folders in favor of PackageReference and PackageDownload items.
- Build all configurations by default when building an individual project (either on the CLI or inside VS) same as with the official SDK. This enables .NETFramework Test Explorer support.
- Centrally define libraries that compose the shared framework instead of in each Directory.Build.props file to be able to build the targeting pack first and consume it in the OOB libraries.
- Use ProjectReferences to reference OOB projects. Compile against the reference assembly but use the implementation assembly app-local during runtime.
- Remove OOBs from the testhost and remove the testhost folder for .NETFramework as it isn't required anymore.
- Only binplace for `$(NetCoreAppCurrent)` to compose a) the targeting pack, b) the runtime pack, c) the testhost, d) a full closure for the shims.
- Use Targeting Packs for OOB projects (with their implicit assembly references) but still explicitly define granular references for .NETCoreApp configurations (`DisableImplicitAssemblyReferences` switch). Use the implicit targeting pack references in some Microsoft.Extensions.* cases.
- Remove placeholder configurations as they aren't needed anymore with explicit P2Ps vs Targeting Pack references.
- Remove implicit assembly references (ie for .NETFramework, mscorlib)
- Remove AssemblySearchPath hacks that were introduced with https://github.com/dotnet/runtime/commit/b7c4cb71fe37e9ec14e97abbcd6cb0f4b4caca84 as the targeting pack is now used by default.
- Reduce unnecessary .NETFramework configurations that were added to run tests in favor of the already existing ref&src configurations.
- Stop hardcoding the paths for wasm assemblies and use the returned TargetPath of the ProjectReferences.
- Addressed formatting (ItemGroups, References at the bottom of the project file, ordering of references, use `LibrariesProjectRoot` instead of a relative path, unnecessary AssemblyName and RootNamespace properties which are identical to the project name, ordering of tfms)

## Commits:
1. adds the necessary infra
2. updates all ref projects to not depend on RefPath
3. updates all source projects to not depend on RefPath
4. updates all test projects to use targeting packs for all configurations and introduces references on top of the targeting pack references
5. adds .NET Framework refs for extensions libs and removes unneeded ones by compiling against the actual product TFM instead of `$(NetFrameworkCurrent)`.
6. reverts https://github.com/dotnet/runtime/commit/535d957421acdf4bff5e8799d14eaf038ae2fa17 as it isn't necessary anymore.

## Post Merge TODO

- [x] https://github.com/dotnet/source-indexer/blob/4448f04f6b279b5a9e816a290febc7fddcb4d2c5/src/index/repositories.props#L12 needs to be updated post merge.
- [x] Update docs for https://github.com/dotnet/runtime/issues/32249
- [x] Update documentation in general for libraries projects

ViktorHofer commented:
@Anipik to avoid collisions with https://github.com/dotnet/runtime/pull/35589 I didn't use Targets* properties in the added code and updated the existing ones where git merge conflicts would happen in your PR.

ericstj commented:
> Without the flag I get ApiCompat errors all over the place. Any idea what could be causing this? Maybe the additional contract path to NetStandard.Library\ref in the nuget cache?

Yep.  Missing dependencies for reference assembly.  You should look at a failure, find what's missing, then find the best way to provide it.   Look here:
https://github.com/dotnet/arcade/blob/5f124ea57f5a2a275ceade393d9a1c2fa1dc1720/src/Microsoft.DotNet.ApiCompat/build/Microsoft.DotNet.ApiCompat.targets#L60-L62
https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/eng/references.targets#L3

Seems these targets assume a single closure complete $(RefPath).  You no longer have that when targeting .NETStandard.  You might be able to just use the dependency directories of the implementation.

ViktorHofer commented:
I see, APICompat is using custom knobs to define an assemblies dependencies.

> You might be able to just use the dependency directories of the implementation.

Can you please elaborate on that? Not sure what you mean by "dependency directories of the implementation".

ericstj commented:
API Compat needs dependency directories for both Contract and Implementation.  APICompat is running in the context of the implementation assembly, so it has access to all the implementation assembly's dependencies via `@(ReferencePath)` items.  It doesn't have any of the contract dependencies, since those are defined in a different project.  In the past it could *assume* that the dependencies of the contract were all present in `$(RefPath)` but that's no longer the case once we stop copying everything.  So I'm imagining a couple paths forward:

1. Call a target in the reference project to get the path to the reference assembly and all it's references.
2. Make an assumption that implementation dependencies can be used as they should be a superset of the reference dependencies and use these for the contract as well.  This assumption will usually hold, but may be different in the case that the contract compiled against a different TFM or type universe (System.Runtime vs netstandard vs System.Private.CoreLib).

ahsonkhan commented on src/libraries/System.Text.Json/src/System.Text.Json.csproj:
```diff
@@ -211,43 +211,55 @@
-    <Reference Include="Microsoft.Bcl.AsyncInterfaces" />
+    <Reference Include="System.Threading.Tasks.Extensions" />
+  </ItemGroup>
+  <ItemGroup Condition="'$(TargetFramework)' == 'netstandard2.0'">
```
Just verifying, are these packaging changes correct and expected, @steveharter, @layomia?

ViktorHofer replied:
These are infra changes only. Instead of bootstrapping and copying these referenced packages upfront we now just PackageReference them.

ericstj replied:
Did you consider making this PR switch everything other than NetCoreAppCurrent to PackageReference?  That might avoid having to add new conditions in most places.

ViktorHofer replied:
I tried that with https://github.com/dotnet/runtime/pull/34893 but it became convoluted. Especially with $(NetFrameworkCurrent) which we run tests against. I'm fine with the extra costs for incremental progress.

ahsonkhan commented on src/libraries/System.Text.Json/src/System.Text.Json.csproj:
```diff
@@ -211,43 +211,55 @@
              Link="Common\System\Collections\Generic\ReferenceEqualityComparer.cs" />
   </ItemGroup>
   <ItemGroup Condition="'$(TargetsNetCoreApp)' == 'true'">
+    <Reference Include="System.Buffers" />
```
Why do we now need to move these into the `"'$(TargetsNetCoreApp)' == 'true'"` condition?

ViktorHofer commented:
> Call a target in the reference project to get the path to the reference assembly and all it's references.

Sounds like the better approach here as we sometimes mix universes between refs and src. Not yet sure how to do that but I will experiment :D

ericstj commented:
> we sometimes mix universes between refs and src

Technically all the dependencies of the ref should be validated independently as we build each of those libraries so it may be safe to use implementation directories for these single-library runs of APICompat.  I agree that calling a target in the ref project might seem cleaner but it may be somewhat expensive for perf.  Check that if you go that route.

ericstj commented on eng/references.targets:
```diff
@@ -1,6 +1,6 @@
 <Project>
   <PropertyGroup>
-    <ContractDependencyPaths>$(RefPath)</ContractDependencyPaths>
+    <ContractDependencyPaths>$(RefPath);@(ReferencePath)</ContractDependencyPaths>
```
I think you need directories, right?

ViktorHofer replied:
sorry don't follow. Are you talking about the item not being available in the property? The item is evaluated inside a target so it is available.

ericstj replied:
`@(ReferencePath)` is the full path to the referenced DLLs, whereas `$(RefPath)` is a directory.  I believe we should be passing the directories, eg `@(ReferencePath->'%(RootDir)%(Directory)')`

ViktorHofer replied:
Got it, will do.

ericstj replied:
(you may need to deduplicate, not sure)

ViktorHofer commented on eng/references.targets:
```diff
@@ -13,21 +20,32 @@
     <_TargetFrameworkDirectories>$(MSBuildThisFileDirectory)</_TargetFrameworkDirectories>
     <_FullFrameworkReferenceAssemblyPaths>$(MSBuildThisFileDirectory)</_FullFrameworkReferenceAssemblyPaths>
```
@ericstj do you know if `_FullFrameworkReferenceAssemblyPaths` and `_TargetFrameworkDirectories` are required for `NetCoreAppCurrent`?

ericstj replied:
This predates the .NET Core SDK when we were using just common/PCL targets to build 1.x.  Those targets had baked in assumptions that every framework would have a machine-wide targeting pack.  I suspect it's not needed any more since the SDK now supports fully package based frameworks.

ViktorHofer reviewed (commented):
Trying to figure out the reason for the remaining errors (libs.src):

```
C:\Users\vihofer\.nuget\packages\microsoft.dotnet.apicompat\5.0.0-beta.20261.9\build\Microsoft.DotNet.ApiCompat.targets(145,5): error : CannotRemoveBaseTypeOrInterface : Type 'System.Text.Json.Utf8JsonWriter' does not implement interface 'System.IAsyncDisposable' in the reference but it does in the implementation. [C:\git\runtime4\src\libraries\System.Text.Json\src\System.Text.Json.csproj]
C:\Users\vihofer\.nuget\packages\microsoft.dotnet.apicompat\5.0.0-beta.20261.9\build\Microsoft.DotNet.ApiCompat.targets(159,5): error : MatchingRefApiCompat failed - The reference assembly doesn't match all the APIs in the implementation for 'C:\git\runtime4\artifacts\bin\System.Text.Json\netcoreapp3.0-Debug\System.Text.Json.dll'. To address either fix errors in the reference assembly (referenced as implementation in compat errors for this reverse compat check), add the issues to the baseline file 'C:\git\runtime4\src\libraries\System.Text.Json\src\MatchingRefApiCompatBaseline.txt' or disable this check by setting RunMatchingRefApiCompat=false in this project. [C:\git\runtime4\src\libraries\System.Text.Json\src\System.Text.Json.csproj]
C:\Users\vihofer\.nuget\packages\microsoft.dotnet.apicompat\5.0.0-beta.20261.9\build\Microsoft.DotNet.ApiCompat.targets(145,5): error : MembersMustExist : Member 'public void System.Runtime.CompilerServices.Unsafe.SkipInit<T>(T)' does not exist in the reference but it does exist in the implementation. [C:\git\runtime4\src\libraries\System.Runtime.CompilerServices.Unsafe\src\System.Runtime.CompilerServices.Unsafe.ilproj]
C:\Users\vihofer\.nuget\packages\microsoft.dotnet.apicompat\5.0.0-beta.20261.9\build\Microsoft.DotNet.ApiCompat.targets(145,5): error : MembersMustExist : Member 'public T System.Runtime.CompilerServices.Unsafe.Unbox<T>(System.Object)' does not exist in the reference but it does exist in the implementation. [C:\git\runtime4\src\libraries\System.Runtime.CompilerServices.Unsafe\src\System.Runtime.CompilerServices.Unsafe.ilproj]
C:\Users\vihofer\.nuget\packages\microsoft.dotnet.apicompat\5.0.0-beta.20261.9\build\Microsoft.DotNet.ApiCompat.targets(159,5): error : MatchingRefApiCompat failed - The reference assembly doesn't match all the APIs in the implementation for 'C:\git\runtime4\artifacts\bin\System.Runtime.CompilerServices.Unsafe\net472-Debug\System.Runtime.CompilerServices.Unsafe.dll'. To address either fix errors in the reference assembly (referenced as implementation in compat errors for this reverse compat check), add the issues to the baseline file 'C:\git\runtime4\src\libraries\System.Runtime.CompilerServices.Unsafe\src\MatchingRefApiCompatBaseline.txt' or disable this check by setting RunMatchingRefApiCompat=false in this project. [C:\git\runtime4\src\libraries\System.Runtime.CompilerServices.Unsafe\src\System.Runtime.CompilerServices.Unsafe.ilproj]
C:\Users\vihofer\.nuget\packages\microsoft.dotnet.apicompat\5.0.0-beta.20261.9\build\Microsoft.DotNet.ApiCompat.targets(145,5): error : CannotRemoveBaseTypeOrInterface : Type 'System.Runtime.CompilerServices.Unsafe' does not inherit from base type 'System.Object' in the reference but it does in the implementation. [C:\git\runtime4\src\libraries\System.Runtime.CompilerServices.Unsafe\src\System.Runtime.CompilerServices.Unsafe.ilproj]
C:\Users\vihofer\.nuget\packages\microsoft.dotnet.apicompat\5.0.0-beta.20261.9\build\Microsoft.DotNet.ApiCompat.targets(159,5): error : MatchingRefApiCompat failed - The reference assembly doesn't match all the APIs in the implementation for 'C:\git\runtime4\artifacts\bin\System.Runtime.CompilerServices.Unsafe\net45-Debug\System.Runtime.CompilerServices.Unsafe.dll'. To address either fix errors in the reference assembly (referenced as implementation in compat errors for this reverse compat check), add the issues to the baseline file 'C:\git\runtime4\src\libraries\System.Runtime.CompilerServices.Unsafe\src\MatchingRefApiCompatBaseline.txt' or disable this check by setting RunMatchingRefApiCompat=false in this project. [C:\git\runtime4\src\libraries\System.Runtime.CompilerServices.Unsafe\src\System.Runtime.CompilerServices.Unsafe.ilproj]
```

ViktorHofer committed:
Revert "fix clean (#33758)"

This reverts commit 535d957421acdf4bff5e8799d14eaf038ae2fa17.

ViktorHofer commented:
@Anipik I'm reverting https://github.com/dotnet/runtime/commit/535d957421acdf4bff5e8799d14eaf038ae2fa17 in this PR as I've fixed the actual issue (missing ref configuration for a given source). Without the revert, ApiCompat fails if the ref project hasn't been built yet, which is plausible now that you don't need to build libs.ref anymore for anything != NetCoreAppCurrent.

ViktorHofer commented:
@ericstj I'm currently using named references for live built dependencies which aren't part of a targeting pack (ie most of Microsoft.Extensions.*). Should I switch entirely to P2Ps for these or keep the named ones? I would prefer P2Ps and I noticed that we already use them in some src projects but wanted to consult you first.

ViktorHofer commented on src/libraries/System.Windows.Extensions/src/System.Windows.Extensions.csproj:
```diff
@@ -56,14 +56,17 @@
   </ItemGroup>
   <ItemGroup>
+    <!-- Need to reference the ref which has an OS agnostic configuration for netcoreapp3.0. -->
+    <ProjectReference Include="..\..\System.Drawing.Common\ref\System.Drawing.Common.csproj" />
```
@ericstj @safern is it expected that the impl of S.D.Common doesn't have an OS agnostic configuration?

ericstj replied:
Looks like it's being harvested: https://github.com/dotnet/corefx/pull/39275/commits/8e9adad67a084834c7888e643fdd1e720a70b8d8

ericstj commented:
Using P2Ps for src -> src increases the depth of the build graph and will thus increase overall build time and dev inner-loop.  It also violates the principal of our reference assemblies being the contract exposed by an assembly.  Using P2Ps from src -> ref would preserve these characteristics.  As we've discussed we can eventually merge the reference assembly build into the src projects as another mode and that might make this behave a bit more naturally.  When we eventually do that, we can probably make the P2P behavior look just like it does with the CSC /refout feature, only we'll have our checked in source file for producing the reference assembly.

A couple other thoughts here:
1. See if you can split this work up into smaller chunks that can be committed independently.  Otherwise it will become a slog.  One thought would be to do what infra work is needed to enable this, do projects in batches, then mop up by deleting the dep-projs after that (once they are no longer used).
2. Consider diffing nuspecs to mitigate accidental changes from this.  I'd expect this change to have no impact on what we're shipping.

Anipik commented on src/libraries/Microsoft.Extensions.Caching.Abstractions/src/Microsoft.Extensions.Caching.Abstractions.csproj:
```diff
@@ -7,7 +7,7 @@
 
   <ItemGroup>
-    <Reference Include="Microsoft.Extensions.Primitives" />
+    <ProjectReference Include="..\..\Microsoft.Extensions.Primitives\ref\Microsoft.Extensions.Primitives.csproj" />
```
We started adding project references to the .sln filesin order to allow building from visual studio, you should add that for th e new project references

Anipik replied:
I think we might not require that as we just build the latest netcoreapp in VS. Can you verify it.

ViktorHofer replied:
> I think we might not require that as we just build the latest netcoreapp in VS. Can you verify it.

With this work we want to build everything inside VS. I will update the slns, thanks.

Anipik commented on src/libraries/Microsoft.Extensions.Configuration.CommandLine/src/Microsoft.Extensions.Configuration.CommandLine.csproj:
```diff
@@ -1,13 +1,18 @@
     <EnableDefaultItems>true</EnableDefaultItems>
   </PropertyGroup>
 
+  <ItemGroup Condition="'$(TargetFramework)' == '$(NetCoreAppCurrent)'">
```
just for curiosity, why we are adding the new dependencies here ?

ViktorHofer replied:
The library wasn't building for .NETCoreApp before correctly because of the `TargetFramework` being set to `$(DefaultNetCoreTargetFramework)` which doesn't exist in this repo. Probably a leftover from the dotnet/extensions migration. Fixing this requires the added references.

Anipik commented on eng/references.targets:
```diff
@@ -1,8 +1,15 @@
     <FrameworkPathOverride>$(RefPath)</FrameworkPathOverride>
-    <AssemblySearchPaths>$(AssemblySearchPaths);$(RefPath);{RawFileName}</AssemblySearchPaths>
+    <IncludeDefaultReferences Condition="'$(IncludeDefaultReferences)' == ''">true</IncludeDefaultReferences>
+    <IncludeNETStandardReference Condition="'$(IncludeDefaultReferences)' == 'true' and
```
why we are not setting IncludeNETStandardReference to true when TFI is netstandard ?

ViktorHofer replied:
References for .NETStandard are coming via the TargetingPack by default, no need to manually reference netstandard.dll.

Anipik commented on eng/references.targets:
```diff
@@ -36,32 +54,41 @@
   </Target>
 
-  <Target Name="RemoveConflictResolutionAssetsForTests"
+  <Target Name="RemoveDefaultReferenceExclusions"
```
when will we be using this target ?

ViktorHofer replied:
That target is already in master: https://github.com/dotnet/runtime/blob/master/eng/references.targets#L57, I just renamed it and modified it slightly.

Its purpose is to remove references that are added from the RefPath.

safern commented on eng/build.ps1:
```diff
@@ -134,6 +134,11 @@ foreach ($argument in $PSBoundParameters.Keys)
 }
 
+# Default the BuildTargetFramework to the latest netcorapp version
+if (!$framwork) {
```
$framwork -> $framework?

safern replied:
Also, this should instead be done in MSBuild right? I thought we already did that.
https://github.com/dotnet/runtime/blob/master/src/libraries/Directory.Build.props#L21

ViktorHofer replied:
It needs to be set in the build entry scripts now to differentiate between a root build and an individual project build which now builds all configurations by default.

safern commented on eng/build.sh:
```diff
@@ -298,6 +300,11 @@ while [[ $# > 0 ]]; do
   esac
 done
 
+# Default the BuildTargetFramework to the latest netcorapp version
```
nit: `netcorapp` -> `netcoreapp`.

safern commented on eng/references.targets:
```diff
@@ -1,33 +1,43 @@
   <PropertyGroup>
-    <ContractDependencyPaths>$(RefPath)</ContractDependencyPaths>
+    <IncludeDefaultReferences Condition="'$(IncludeDefaultReferences)' == ''">true</IncludeDefaultReferences>
+    <IncludeNETStandardReference Condition="'$(IncludeDefaultReferences)' == 'true' and
```
Why do we need to include netstandard when TFI != .NETStandard?

ViktorHofer replied:
That was added by @weshaggard here: https://github.com/dotnet/buildtools/pull/1460/files#diff-8f8878b439291dd5d102ff6d2414ce67R59-R62. AFAIK this is to combine different type universes (System.Runtime vs netstandard vs mscorlib). cc @ericstj

joperezr replied:
Yeah, if we want to allow implementations that target netcoreapp to be able to reference netstandard assemblies then we need the netstandard facades.

ViktorHofer commented:
Please note that I'm still in a WIP state. I will probably continue push forcing and separating my commits by concerns at the end this is ready so that the changes are reviewable.

safern commented on eng/references.targets:
```diff
@@ -1,33 +1,43 @@
+                 Private="false" />
+      <!-- Filter out experimental reference assemblies. -->
+      <Reference Remove="@(Reference)"
+                 Condition="$([System.String]::new('%(FileName)%(Extension)').Contains('Experimental'))" />
```
Has this been like that always?

ViktorHofer replied:
For Utf8String yes: https://github.com/dotnet/runtime/blob/e0e752ac9ee67f3dc7fa127d6f7495132b4bcee7/src/libraries/Directory.Build.props#L358-L365.

safern commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/Microsoft.Extensions.Caching.Memory.csproj:
```diff
@@ -7,11 +7,11 @@
-    <Reference Include="Microsoft.Extensions.Logging.Abstractions" />
-    <Reference Include="Microsoft.Extensions.Options" />
-    <Reference Include="Microsoft.Extensions.Primitives" />
+    <ProjectReference Include="..\..\Microsoft.Extensions.Caching.Abstractions\ref\Microsoft.Extensions.Caching.Abstractions.csproj" />
```
So this means that every time we add a new reference we will need to update the SLN?

ViktorHofer replied:
If we continue using our current approach (having distinct solution files per leaf) then yes. I'm experimenting with using solution filter files instead (.slnf) and having a central solution file in the libraries root.

safern commented on src/libraries/Microsoft.Extensions.Configuration.Json/src/Microsoft.Extensions.Configuration.Json.csproj:
```diff
@@ -8,18 +8,22 @@
-    <Reference Include="Microsoft.Extensions.FileProviders.Abstractions" />
-    <Reference Include="System.Memory" />
-    <Reference Include="System.Text.Json" />
+    <ProjectReference Include="..\..\Microsoft.Extensions.Configuration\ref\Microsoft.Extensions.Configuration.csproj" />
```
Should we introduce a `LibrariesRootDir` or something like that to make this simpler? IIRC there was already one.

ViktorHofer replied:
I prefer the relative path to be honest. Relying on an msbuild property a) makes the project file harder to read and b) makes it less portable.

joperezr replied:
That really depends on which name you use for the property, but if you choose correctly I would argue that it is much better because of the exact two reasons you pointed above. Relative is less readable as I need to know the current directory, and it is less portable in case we want to add any folders in between.

ViktorHofer replied:
Giving it a second thought, I agree. Mainly because the existing property `LibrariesProjectRoot` is unambiguous and conveys the meaning well. I used that variable in places that I touched and where it made sense (referencing another leaf project). Thanks

safern commented on src/libraries/Microsoft.Extensions.Configuration.Json/src/Microsoft.Extensions.Configuration.Json.csproj:
```diff
@@ -8,18 +8,22 @@
   <ItemGroup Condition="'$(TargetFramework)' == '$(NetCoreAppCurrent)'">
     <Reference Include="System.Collections" />
     <Reference Include="System.Linq" />
+    <Reference Include="System.Memory" />
```
I'm confused, when do we need to use P2P to ref assemblies and when should we not?

safern replied:
Also, if the end goal is to be able to build an individual project without an upfront build, this won't work for `NetCoreAppCurrent` because we haven't built these ref assemblies, correct?

ViktorHofer replied:
My goal is to use P2Ps everywhere (which is challenging because of all libraries being evaluated which the dependency chain) but that's out of scope for this PR. I added/changed to P2Ps where non NetCoreAppCurrent configurations were referencing that project.

joperezr commented on eng/Versions.props:
```diff
@@ -70,7 +71,38 @@
     <MicrosoftNETCoreILAsmVersion>5.0.0-preview.4.20202.18</MicrosoftNETCoreILAsmVersion>
     <MicrosoftNETSdkILVersion>5.0.0-preview.4.20202.18</MicrosoftNETSdkILVersion>
     <!-- Libraries dependencies -->
+    <SystemBuffersVersion>4.5.1</SystemBuffersVersion>
```
How are we ensuring that these versions are kept up-to-date? Are we using darc? If so, what if some packages are only serviced out of a different branch, for example release/2.1? How will we get an update?

ViktorHofer replied:
For 3.1 we will use BAR and Darc. That isn't possible for 2.1 unfortunately so we need to update these versions manually. Nothing changes here, these versions were already hardcoded in the depproj restore: https://github.com/dotnet/runtime/blob/e0e752ac9ee67f3dc7fa127d6f7495132b4bcee7/src/libraries/restore/binplacePackages/binplacePackages.depproj#L31-L42. I'm just moving them into Versions.props.

FWIW we could consider using something like dependabot.

ViktorHofer committed:
Remove depprojs in favor of PackageReferences

Remove depprojs which currently binplace external references into the
RefPath folders in favor of PackageReference or PackageDownload items.
Use targeting packs and prefer live built assemblies with P2Ps and
simple name references.

Use implementation dependencies in RefApiCompat

Currently RunMatchingRefApiCompat expects a single dependency closure
directory which isn't true anymore with this change. When validating API
compatibility for a matching reference assembly, we run as part of the
implementation assembly. Therefore adding the ReferencePath items to the
ContractDepenencyPaths to guarantee full closure via the
implementation's dependencies.

Build for all configurations by default when not building from root
(i.e. dotnet build x.csproj) as expected from the SDK.

Build the shared framework and the ref and runtime packs first and make
out-of-band projects build on top of the targeting packs (for net5.0
as well).

Remove ReferenceFromRuntime items and replace with P2Ps.

Remove the concept of Ref- and RuntimePath and use the locally built
refpack and runtimepack for net5.0 OOBs and shims.

ViktorHofer committed:
Update ref projects to not depend on RefPath

ViktorHofer committed:
Update src projs to stop using RefPath for OOBs

ViktorHofer committed:
Use targeting packs in test projects

Rely on targeting packs for all configurations. Use the locally built
refpack for NetCoreAppCurrent.

ViktorHofer commented:
/azip run runtime

ViktorHofer commented:
/azp run runtime

ViktorHofer commented:
/azp run runtime

ViktorHofer commented:
/azp run runtime

ViktorHofer commented:
/azp run runtime

ViktorHofer committed:
Remove unneeded .NETFramework configurations

ericstj commented on src/libraries/System.Resources.Extensions/tests/System.Resources.Extensions.Tests.csproj:
```diff
@@ -16,7 +16,15 @@
+    <ProjectReference Include="..\src\System.Resources.Extensions.csproj" />
   </ItemGroup>
+  <ItemGroup Condition="$(TargetFramework.StartsWith('$(NetCoreAppCurrent)'))">
+    <ProjectReference Include="$(LibrariesProjectRoot)System.Drawing.Common\src\System.Drawing.Common.csproj" />
```
You'll also need System.Windows.Extensions to get the drawing TypeConverters.

Can you also double-check that you can run the GenerateTestResourcesFile target below and it doesn't change the .resources file?

ViktorHofer replied:
Thanks, I wasn't able to figure that out :(

> Can you also double-check that you can run the GenerateTestResourcesFile target below and it doesn't change the .resources file?

Already did yesterday. The target doesn't work in it's current state because `TargetDir` and `TargetFile` aren't set but I hardcoded the paths and the file didn't change.

ViktorHofer commented on src/libraries/NetCoreAppLibrary.props:
```diff
@@ -0,0 +1,160 @@
+<Project>
+  <PropertyGroup>
+    <NetCoreAppLibrary>
+      Microsoft.CSharp;
```
Discussed this with @ericstj offline. We need a way to determine which library is part of the shared framework so that we can build the targeting pack first and later consume it in OOB libraries. There were at least two options:
a) add a target that returns the value which is called by src.proj/ref.proj/tests.proj. Big downside is cost and potential static graph restore issues.
b) Centralize the list of libraries that compose the shared framework (approach that I took). Downside: it's not defined per leaf anymore but in favor no additional build costs.

safern replied:
I see... I think it helps with having a centralized list of what's inbox (it helps and sometimes people struggle with finding what's inbox and what's not). 

I guess the reason why we chose a Property instead of an Item (which is more readable), was for simplicity of figuring out if a project was inbox?

ViktorHofer replied:
> I see... I think it helps with having a centralized list of what's inbox (it helps and sometimes people struggle with finding what's inbox and what's not).

Jep, agree.

> I guess the reason why we chose a Property instead of an Item (which is more readable), was for simplicity of figuring out if a project was inbox?

I actually started with an Item but had to change to a property as other properties relied on IsNetCoreApp.

safern replied:
I see, ok. Sounds good. Thanks for explaining the motivation of changing this 😄

ViktorHofer commented:
@Anipik @ericstj @joperezr @safern finally the PR should be ready and I already reviewed it myself. I summarized the changes in the top post with a list of the commits. I already merged some related changes into master and arcade in the last weeks. The remaining changes in this PR weren't easy to encapsulate further.

I highly recommend to review per commit. Will share out perf numbers with best and worst case build times later after CI has completed.

ViktorHofer commented:
@akoeplinger @safern I don't understand the failures in the WASM leg, maybe you can help out? Would like to get this merged in the next days :)

ericstj commented on eng/AvoidRestoreCycleOnSelfReference.targets:
```diff
@@ -0,0 +1,14 @@
+<Project>
+  <!-- Set PackageId to a temporary value to avoid NuGet restore cycle error NU1108: https://github.com/NuGet/Home/issues/6754 -->
+  <PropertyGroup>
+    <PackageIdTemp>$(PackageId)</PackageIdTemp>
```
nit: should this be underscore property?

ericstj commented on eng/CapturedBuildOutputCrossTargetingTargets.targets:
```diff
@@ -0,0 +1,6 @@
+<!-- Overriding the common Build target to capture the inner outputs: https://github.com/microsoft/msbuild/issues/2148. -->
```
It looks like this issue is closed, should we reopen it?

ViktorHofer replied:
Asked @rainersigwald to reopen: https://github.com/microsoft/msbuild/issues/2148

ericstj commented on eng/build.ps1:
```diff
@@ -65,7 +65,7 @@ function Get-Help() {
   Write-Host "  -allconfigurations      Build packages for all build configurations."
   Write-Host "  -coverage               Collect code coverage when testing."
-  Write-Host "  -framework (-f)         Build framework: net5.0 or net472."
+  Write-Host "  -framework (-f)         Build framework: net5.0, net48 or any other TFM."
```
Would we really support some other TFM being passed?  I can't imagine we do the right thing when someone passes netstandard2.1 here for instance.

ViktorHofer replied:
If you pass in netstandard2.1 you would build the best configuration if there's one for the projects. Meaning you would build netstandard2.1 and netstandard2.0 refs and sources. That already works today, or isn't that what you would expect to happen?

ericstj replied:
That's what it does, but is it meaningful?  Would someone ever want to do this?  The parts that I was imagining wouldn't do the right thing are shims, pretest, any pkgs.

ViktorHofer replied:
shims and pretest are conditioned to only run on `$(NetCoreAppCurrent)`. Unsure about packages. For the purprose of this PR I will remove the "any other TFM" note.

ericstj commented on eng/build.ps1:
```diff
@@ -190,6 +190,11 @@ foreach ($argument in $PSBoundParameters.Keys)
 
+# Default the BuildTargetFramework to the latest netcorapp version
+if (!$framework) {
+  $arguments += " /p:BuildTargetFramework=net5.0"
```
Why do we need to do this?  It will cause things to behave slightly differently when building from root (global property) vs leaf (default value from props)

ViktorHofer replied:
When building a leaf we now build all configurations by default. The TargetFramework SDK doesn't filter anymore based on the `BuildAllConfigurations` property but based on the `BuildTargetFramework` property. `BuildTargetFramework` isn't set to a default value anymore.

safern replied:
Why aren't we setting it to a default anymore?

ViktorHofer replied:
As we now sequence all dependencies into the projects instead of doing an upfront depproj restore, we don't need to filter the `TargetFrameworks` to `$(NetCoreAppCurrent)` anymore and can build all configurations instead which is the same behavior as in a customer application (defined by the sdk).

ericstj replied:
> When building a leaf we now build all configurations by default. 

I see, this makes sense.

safern replied:
Can we instead set `BuildTargetFramework` in `Build.proj` which is the one that comes into play when we're not building a leaf project?

ViktorHofer replied:
That's a great suggestion, let me do that as it will mean that `dotnet build` will continue to work from the repo root.

ViktorHofer replied:
At some point we should move all global properties being passed in from build scripts into Build.proj.

ericstj commented on eng/referenceAssemblies.props:
```diff
@@ -1,9 +1,9 @@
+                            !$(BuildTargetFramework.StartsWith('netstandard')) and
+                            !$(BuildTargetFramework.StartsWith('net4'))">
+    <AdditionalBuildTargetFrameworks>$(AdditionalBuildTargetFrameworks);netstandard2.0</AdditionalBuildTargetFrameworks>
+    <AdditionalBuildTargetFrameworks Condition="!$(BuildTargetFramework.StartsWith('netcoreapp2.0'))">$(AdditionalBuildTargetFrameworks);netstandard2.1</AdditionalBuildTargetFrameworks>
```
When would we ever specify `BuildTargetFramework` as `netcoreapp2.0`?  I can imagine cases where `TargetFramework` is this, but not `BuildTargetFramework`.  That implies someone would be trying to build a netcoreapp2.0 vertical from the latest codebase 😕

ViktorHofer replied:
> That implies someone would be trying to build a netcoreapp2.0 vertical from the latest codebase 😕

Right, you can already do that today. This change just fixes the over building of the reference assembly configurations for some .NETCoreApp tfms.

ericstj replied:
It still doesn't make sense to me.  I don't understand the purpose of building a `netcoreapp2.0` vertical.  Additionally,  `netcoreapp2.0` doesn't even support `netstandard2.1`  why would we be building that when someone wants to build `netcoreapp2.0`

ViktorHofer replied:
> Additionally, netcoreapp2.0 doesn't even support netstandard2.1 why would we be building that when someone wants to build netcoreapp2.0

That's why the condition is inverted. We only add the netstandard2.1 ref assembly addition build target framework for everything .NETCoreApp except 2.0.

> It still doesn't make sense to me. I don't understand the purpose of building a netcoreapp2.0 vertical.

As said, this already works in master today and I'm only fixing the overbuilding here. If we want to stop supporting other BuildTargetFrameworks, let's do that in a follow-up PR.

ericstj replied:
Never mind, this was me missing a `!`  sorry for the noise.

ericstj commented on eng/references.targets:
```diff
@@ -1,67 +1,52 @@
+          AfterTargets="AssignProjectConfiguration"
+          BeforeTargets="PrepareProjectReferences"
+          Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp' and
+                     ('$(IsTestProject)' == 'true' or '$(IsTestSupportProject)' == 'true') and
```
Is the `IsTest*` condition necessary?

ViktorHofer replied:
P2Ps are private by default for ref and src projects which makes this target a no-op for such.

ericstj commented on eng/resolveContract.targets:
```diff
@@ -1,17 +1,95 @@
   <!-- intentionally empty since we no longer need a target -->
   <Target Name="ResolveMatchingContract" />
+  
+  <Target Name="SetILResourceReference"
```
We want this to be centrally defined?  It's very specific to the ILProj hack. https://github.com/dotnet/runtime/commit/61178c65bd919ae8d82eade1797ff468126a948c

ViktorHofer replied:
I wasn't sure myself. Would you prefer having it in the single ilproj that we currently have?

ericstj replied:
I think so, or at least it needs some comments around its purpose.

ViktorHofer replied:
moved.

ericstj commented on eng/resolveContract.targets:
```diff
@@ -1,17 +1,95 @@
+          BeforeTargets="RunMatchingRefApiCompat"
+          Condition="'$(HasMatchingContract)' == 'true' and '$(ContractTargetFramework)' != ''">
+    <MSBuild Projects="$(ContractProject)"
+             Properties="TargetFramework=$(ContractTargetFramework);Configuration=$(Configuration);BuildProjectReferences=false"
```
Due to global properties, this will result in another evaluation of the referenced project.   If I follow the logic here we have two cases.
1.  ContractAssemblyPath exists, in this case it needs to be closure complete, so we can just use this directory for references.
2. ContractAssemblyPath doesn't exist.  We need to add a ProjectReference to ContractProject.  We should be able to return the ReferencePathDirectories on this item without another call to the project.

ViktorHofer replied:
Thanks, I removed the logic entirely as there is no need for such a target anymore.

ericstj commented on eng/targetingpacks.targets:
```diff
@@ -0,0 +1,110 @@
+          Condition="'$(_UseLocalTargetingRuntimePack)' == 'true'"
+          AfterTargets="ResolveFrameworkReferences">
+    <ItemGroup>
+      <ResolvedTargetingPack Update="@(ResolvedTargetingPack)"
```
I believe `ResolvedTargetingPack Update="@(ResolvedTargetingPack)"` is redundant.  In a target if you omit include/update/remove you are operating on all items.

ViktorHofer replied:
@rainersigwald just to double check, is this indeed true?

ericstj replied:
Docs : https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-items?view=vs-2019#updating-metadata-on-items-in-an-itemgroup-of-a-target

ViktorHofer replied:
I see, so this can only be omitted inside a Target. Thanks

ericstj commented on eng/testing/netfx.exe.config:
```diff
@@ -1,7 +1,6 @@
 <?xml version="1.0" encoding="utf-8"?>
 <configuration>
   <runtime>
-    <developmentMode developerInstallation="true" />
```
Interesting, so we no longer rely on DEVPath and we get all the right bindingRedirects?  That's significant improvement to test coverage.

ViktorHofer replied:
Exactly. There was an issue with them being overwritten but I fixed that.

ericstj commented on src/libraries/Directory.Build.props:
```diff
@@ -3,21 +3,20 @@
   <PropertyGroup>
     <BeforeTargetFrameworkInferenceTargets>$(RepositoryEngineeringDir)BeforeTargetFrameworkInference.targets</BeforeTargetFrameworkInferenceTargets>
-    <ToolSetCommonDirectory>$(RepoRoot)artifacts\toolset\Common\</ToolSetCommonDirectory>
+    <CustomAfterMicrosoftCommonCrossTargetingTargets>$(RepositoryEngineeringDir)CapturedBuildOutputCrossTargetingTargets.targets</CustomAfterMicrosoftCommonCrossTargetingTargets>
```
Would it make sense to do this in our TargetFramework SDK?

ViktorHofer replied:
I don't think this should go into our TargetFramework.Sdk as it isn't related to multi-targeting by runtime.

ericstj replied:
Fair, if we think MSBuild will fix it then I'm not too concerned.

ericstj commented on src/libraries/Directory.Build.props:
```diff
@@ -66,6 +68,13 @@
 
+  <!-- Project references shouldn't be copied to the output for non test apps. -->
+  <ItemDefinitionGroup Condition="'$(IsTestProject)' != 'true' and '$(IsTestSupportProject)' != 'true'">
+    <ProjectReference>
```
Would it make sense to put this next to `UpdateProjectReferencesWithPrivateAttribute`?

ericstj reviewed (commented):
Some initial feedback after reviewing up to libraries projects.

safern commented on eng/AvoidRestoreCycleOnSelfReference.targets:
```diff
@@ -0,0 +1,14 @@
+<Project>
+  <!-- Set PackageId to a temporary value to avoid NuGet restore cycle error NU1108: https://github.com/NuGet/Home/issues/6754 -->
```
Where were we hitting this?

ViktorHofer replied:
in DiagnosticSource which is an implicit package of NetStandard.Library. Discussed this with @ericstj and based on the NuGet issue this seems to be the cleanest approach for now. You can see that I set that new property in both the ref and src DiagnosticSource projects.

safern commented on eng/pipelines/coreclr/templates/perf-job.yml:
```diff
@@ -91,10 +91,12 @@ jobs:
 
-    - script: "build.cmd -subset libs.pretest -configuration release -ci -arch $(archType) -testscope innerloop /p:RuntimeArtifactsPath=$(librariesDownloadDir)\\bin\\mono\\$(osGroup).$(archType).$(buildConfigUpper) /p:RuntimeFlavor=mono;xcopy $(Build.SourcesDirectory)\\artifacts\\bin\\testhost\\$(_Framework)-$(osGroup)-$(buildConfigUpper)-$(archType)\\* $(Build.SourcesDirectory)\\.dotnet-mono /E  ...
+      # Copy the runtime directory into the testhost folder to include OOBs.
+
```
Is this extra line intentional?

ViktorHofer replied:
yes, as it applies to both statements below :)

safern commented on eng/references.targets:
```diff
@@ -1,67 +1,58 @@
-      </Reference>
-    </ItemGroup>
-  </Target>
+  <!-- netstandard is a default reference whenever building an implementation assembly for .NETCoreApp. -->
```
Why is this? This was not the behavior we had, was it?

ViktorHofer replied:
Before we added the reference to both .NETFramework and .NETCoreApp source projects but with @joperezr recent work this isn't necessary anymore and actually shouldn't be done anymore so that the netfx libraries work without the .NETStandard shims.

safern commented on eng/references.targets:
```diff
@@ -1,67 +1,58 @@
-    <ContractDependencyPaths>$(RefPath)</ContractDependencyPaths>
-    <FrameworkPathOverride>$(RefPath)</FrameworkPathOverride>
-    <AssemblySearchPaths>$(AssemblySearchPaths);$(RefPath);{RawFileName}</AssemblySearchPaths>
+    <IncludeDefaultReferences Condition="'$(IncludeDefaultReferences)' == ''">true</IncludeDefaultReferences>
```
Do we want to include default references for `NETCoreAppCurrent`?

safern replied:
I see we need it in order to use the ref and runtime pack.

ViktorHofer replied:
`IncludeDefaultReferences` is a historical property that only defines if the netstandard reference should be added. There are only very few places that use that property: https://github.com/dotnet/runtime/search?q=IncludeDefaultReferences&unscoped_q=IncludeDefaultReferences. The shims wouldn't event need that property as `IsSourceProject` isn't true for them.

I might clean this up as part of this PR if I have time.

ViktorHofer replied:
Looks like you added that property: https://github.com/dotnet/runtime/commit/6ec72ff7ea729936438de20cb95ecba8e710e9d9#diff-d3ed5cb795a36ae602695c7d88edbe72R7

ViktorHofer replied:
Ok deleted that code as the property isn't needed anymore. Thanks.

safern commented on eng/restore/repoRestore.props:
```diff
@@ -1,5 +1,6 @@
 <Project TreatAsLocalProperty="ExcludeRestorePackageImports">
   <PropertyGroup>
+    <ToolSetCommonDirectory>$(RepoRoot)artifacts\toolset\Common\</ToolSetCommonDirectory>
```
NormalizePath?

ViktorHofer replied:
I just moved that property from Directory.Build.props into this file as it is only needed for projects that don't support restore (the remaining runtime.depproj and the pkgprojs). No need to change it.

safern commented on eng/targetingpacks.targets:
```diff
@@ -0,0 +1,106 @@
+                         DesignTimeResolveAssemblyReferences"
+          Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp'">
+    <PropertyGroup>
+      <AssemblySearchPaths>$(AssemblySearchPaths);$(MicrosoftNetCoreAppRefPackRefDir.TrimEnd('/\'))</AssemblySearchPaths>
```
We do we need this is we're already updating the ResolvedTargetingPack and RuntimePack?

ViktorHofer replied:
The AssemblySearchPath isn't set by the targetingpack and hence we need to set it manually. As we support both the targetingpack (>3.x) and the packagereference model (2.x) there is no common way when to set it so I sequenced it before it is used in `ResolveAssemblyReferences` and `DesignTimeResolveAssemblyReferences`.

safern commented on eng/targetingpacks.targets:
```diff
@@ -0,0 +1,106 @@
+    <ItemGroup>
+      <ResolvedTargetingPack Path="$(MicrosoftNetCoreAppRefPackDir.TrimEnd('/\'))"
+                             NuGetPackageVersion="$(ProductVersion)"
+                             PackageDirectory="$(MicrosoftNetCoreAppRefPackDir.TrimEnd('/\'))"
```
Why do we need to trimEnd here? Why not just TrimEnd in the property declaration instead to avoid all this TrimEnd duplication?

ViktorHofer replied:
For consistency with our other properties that point to a directory. Most of the time we use the NormalizeDirectory function and if not then just because of historical reasons. You would expect that the property ends with a slash. In these cases I'm trimming the trailing slash as that's the format the sdk expects. It would probably not cause any issues with the slash but I would need to look away :D

safern commented on eng/testing/.runsettings:
```diff
@@ -18,10 +18,6 @@
     <DotNetHostPath>$$DOTNETHOSTPATH$$</DotNetHostPath>
-    <EnvironmentVariables>
-      <!-- Use our self-built framework on .NET Framework -->
-      <DEVPATH>$$DEVPATH$$</DEVPATH>
```
Why is this not needed anymore? We do have some implementations that we built and that are not on the GAC.

ViktorHofer replied:
This isn't needed anymore as all .NETFramework libraries are app-local now. There is now need for the additional runtime directory anymore. Assembly binding redirects handle the cases where we target a live build asset that is present in the framework and will prefer ours same as in a customer app. This means that we now test the same way our customers do which is a nice plus.

safern commented on eng/testing/runtimeConfiguration.targets:
```diff
@@ -1,45 +1,31 @@
+    <GenerateAppConfigurationFile Condition="'$(IsTestProjectOrExecutable)' == 'true' and
+                                             '$(TargetFrameworkIdentifier)' == '.NETFramework'">true</GenerateAppConfigurationFile>
+    <AppConfig Condition="'$(GenerateAppConfigurationFile)' == 'true'">$(MSBuildThisFileDirectory)netfx.exe.config</AppConfig>
+    <!-- TODO: Change in RemoteExecutor -->
```
Open an issue?

ViktorHofer replied:
No need for that, I won't forget and will submit the change soon after this PR is in. The extra work for having a tracking issue isn't worth it.

safern commented on eng/testing/tests.props:
```diff
@@ -29,8 +29,6 @@
     <RuntimeIdentifier>$(PackageRID)</RuntimeIdentifier>
     <SelfContained>true</SelfContained>
-    <EnableTargetingPackDownload>false</EnableTargetingPackDownload>
-    <PlatformManifestFile />
```
Is PlatformManifestFile no longer set to a custom one?

ViktorHofer replied:
No we don't set PlatformManifestFile anymore ourselves anywhere as we now use the live built TargetingPack which contains the PlatformManifest file and the SDK handles all the plumbing, same as in a customer app :)

safern commented on eng/testing/tests.targets:
```diff
@@ -50,6 +45,9 @@
                   Overwrite="true" />
+    
+    <!-- Save disk space. -->
+    <RemoveDir Condition="'$(TargetOS)' == 'Browser'" Directories="$(BundleDir)" />
```
This is handled by @akoeplinger PR so we will need to remove this.

safern commented on src/libraries/Directory.Build.targets:
```diff
@@ -17,6 +14,22 @@
+                                $(NetCoreAppLibrary.Contains('$(AssemblyName);'))">true</IsNETCoreAppSrc>
+    <IsNETCoreAppRef Condition="'$(IsNETCoreAppRef)' == '' and
+                                '$(IsReferenceAssembly)' == 'true' and
+                                $(NetCoreAppLibrary.Contains('$(AssemblyName);')) and
```
Why do we need the `;` at the end?

ViktorHofer replied:
as without the ; the contains check would match a library that isn't the exact match, ie `System.ComponentModel` in `System.ComponentModel.Composition`. Including the semicolon guarantees that we only accept exact matches.

akoeplinger replied:
the only issue is when someone adds another library at the end of the `NetCoreAppLibrary` list and forgets the trailing semicolon, but that is probably rare :)

safern replied:
Ah right... didn't think about that case, the reason why I was asking was the same concern as @akoeplinger mentioned.

ViktorHofer replied:
yeah you are right. #fixed (will push soon).

ViktorHofer replied:
ok undid my change as it caused problems as there was a whitespace and newline now between the last entry and the added semicolon. So I'm fine with the current approach as CI just proved that tests will fail otherwise to build as a default reference is then missing.

safern commented on src/libraries/Microsoft.Extensions.Caching.Memory/tests/Microsoft.Extensions.Caching.Memory.Tests.csproj:
```diff
@@ -10,7 +10,11 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\src\Microsoft.Extensions.Caching.Memory.csproj" SkipUseReferenceAssembly="true" />
```
What does `SkipUseReferenceAssembly` do and why is it needed only here? I thought referencing the src project was like a `ReferenceFromRuntime`?

ViktorHofer replied:
The new model is that P2Ps always target the src projects (except for ref --> ref of course) but compile against the reference assembly that is produced in the ref project. This mimics the RefOut feature as you get the benefit of compiling against the ref assembly but get the src assembly as an app-local asset into the test app's output directory. Because of this we now don't need to store OOB assemblies in the testhost folder anymore and consume them the same way as customers do (app-local).

The `SkipUseReferenceAssembly` allows to compile against the src assembly which is the replacement for `ReferenceFromRuntime` but based on a P2P instead of a simple name reference.

safern commented on src/libraries/Microsoft.Extensions.Configuration.CommandLine/src/Microsoft.Extensions.Configuration.CommandLine.csproj:
```diff
@@ -1,18 +1,15 @@
-    <ExcludeCurrentFullFrameworkFromPackage>true</ExcludeCurrentFullFrameworkFromPackage>
+    <TargetFrameworks>$(NetCoreAppCurrent);netstandard2.0;net461</TargetFrameworks>
     <EnableDefaultItems>true</EnableDefaultItems>
+    <!-- Use targeting pack references instead of granular ones in the project file. -->
```
what do you mean by this, I didn't get what are you referring to with `granular`.

ViktorHofer replied:
This means that we use the default targeting pack references (all) for this project instead of listing the individual references in the project file. We don't want to do this for libs that are part of the shared framework or for OOBs where dependencies matter but for some of the Microsoft.Extensions projects it doesn't matter. I'm experimenting with this feature in these few projects.

safern commented:
Looked at the wasm test failures it seems like this change is causing to no longer use the live built runtime pack when we publish. 

Looking at this offline with @ViktorHofer

Anipik commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/Microsoft.Extensions.Caching.Memory.csproj:
```diff
@@ -7,16 +7,11 @@
-    <Reference Include="mscorlib" />
-    <Reference Include="System.Core" />
+    <ProjectReference Include="$(LibrariesProjectRoot)Microsoft.Extensions.Caching.Abstractions\src\Microsoft.Extensions.Caching.Abstractions.csproj" />
+    <ProjectReference Include="$(LibrariesProjectRoot)Microsoft.Extensions.DependencyInjection.Abstractions\src\Microsoft.Extensions.DependencyInjection.Abstractions.csproj" />
```
@ViktorHofer can you add these project references to the .sln files as well ?

ViktorHofer replied:
I would need to do that for the ref ones as well which results in an explosion of each solution file. The workaround is to make sure that you do a CLI restore first (which is indicated by the error in VS) which also works around the other NuGet restore issue that we are currently tracking. So for the sake of this PR I wouldn't want to add them and in a follow-up convert all slns to slnfs and have a root solution file.

ViktorHofer replied:
When using the -vs switch, the CLI restore is incremental so that's covered.

Anipik commented on src/libraries/Microsoft.Win32.SystemEvents/tests/Microsoft.Win32.SystemEvents.Tests.csproj:
```diff
@@ -1,6 +1,6 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks>$(NetCoreAppCurrent)-Windows_NT;$(NetFrameworkCurrent)-Windows_NT</TargetFrameworks>
+    <TargetFrameworks>$(NetCoreAppCurrent)-Windows_NT;net461-Windows_NT</TargetFrameworks>
```
why are we building net461 in favor of net472 ?

safern replied:
This is to match the src project tfms, but we will still be running the tests against the latest net4* installed in the machine.

ViktorHofer commented:
> Looked at the wasm test failures it seems like this change is causing to no longer use the live built runtime pack when we publish.

Checked locally on Ubuntu when building for Browser, publish is using the local runtime pack:

![image](https://user-images.githubusercontent.com/7412651/88278029-006c6380-cce2-11ea-965d-c715d079882b.png)

ViktorHofer commented:
@safern found the reason for the WASM failures :)

> Looked at the wasm test failures it seems like this change is causing to no longer use the live built runtime pack when we publish.

You were right that some assets were overwriting the live built ones as observable in the binlog. The reason for that was that the targeting pack's PlatformManifest file, which is used in the `ResolvePackageFileConflicts` task to handle conflicts between incoming package assets and live built assets, wasn't generated before the mobile test runners were built and hence build dependencies were copied app-local into the test runner's output directories and then overwriting the live built assemblies in the  bundle. Fixed that 😄

ViktorHofer commented:
Queued an official build: https://dnceng.visualstudio.com/internal/_build/results?buildId=743027&view=results

ViktorHofer commented:
Not merging yet but posting the failures:
- WASM: https://github.com/dotnet/runtime/issues/39473
- Libraries Test Run coreclr & runtime-live-build: https://github.com/dotnet/core-eng/issues/9448

EgorBo commented on src/libraries/src.proj:
```diff
@@ -2,41 +2,48 @@
+                      Condition="'$(MSBuildRestoreSessionId)' != ''" />
+  </ItemGroup>
 
-  <Target Name="NativeBinPlace"
```
https://github.com/dotnet/runtime/blob/master/docs/workflow/building/libraries/webassembly-instructions.md#building-the-webassembly-runtime-files -- this task is not unused

ViktorHofer replied:
I'm not removing it, only moving it to where it belongs. Native binplace is now invoked after the native build in build-native.proj and the corresponding wasm files are moved into wasm.targets.

ViktorHofer replied:
I didn't know that this was documented
 I'll send a PR to update the doc.

ViktorHofer replied:
https://github.com/dotnet/runtime/pull/39891
